Issue 14

whyamiroot opened the pull request:
Implemented RecognitionTaskQueue, which uses BlockingCollection methods TryAdd and TryTake to Enqueue and Dequeue RecognitionTasks.
Needs review

whyamiroot committed:
RecognitionTask model

RecognitionTask model to send to analytics service

whyamiroot committed:
Intermediate commit for issue #14

whyamiroot committed:
Merge branch 'master' into issue-14

whyamiroot committed:
Implemented RecognitionTaskQueue - Issue #14

Implemented RecognitionTaskQueue based on BlockingCollection. Dequeue
method uses TryTake method from BlockingCollection; Enqueue method uses
TryAdd method.

whyamiroot commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -1,97 +1,51 @@
-                }
-            }
+            if (instance == null)
+                instance = new RecognitionTaskQueue(new LoggerFactory());
```
Я не уверен, что нужно именно так получать логгер

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/Models/RecognitionTask.cs:
```diff
@@ -0,0 +1,16 @@
+    public class RecognitionTask
+    {
+        [DataMember(Name = "photo")]
+        public string Photo { get; set; }
```
Это же не сама Photo, а только PhotoId?

whyamiroot replied:
Да, просто я называл поле так же, как в API

alex-ks replied:
Это не совсем одно и то же. В API мы посылаем саму фото в Base64, но в RecognitionTask у нас лежит только её id.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/Models/RecognitionTask.cs:
```diff
@@ -0,0 +1,16 @@
+        [DataMember(Name = "photo")]
+        public string Photo { get; set; }
+        [DataMember(Name = "sessionUID")]
+        public string SessionUid { get; set; }
```
Если я правильно помню, то в задаче на распознавание должно быть свойство ModelName, а не SessionUid.

whyamiroot replied:
Если я правильно понимаю, то сами задания на обработку это вот этот [запрос](https://github.com/alex-ks/backend-comptech/wiki/%D0%9E%D0%BF%D0%B8%D1%81%D0%B0%D0%BD%D0%B8%D0%B5-Service-API#%D0%97%D0%B0%D0%BF%D1%80%D0%BE%D1%81-%D0%BD%D0%B0-%D0%BE%D0%B1%D1%80%D0%B0%D0%B1%D0%BE%D1%82%D0%BA%D1%83-%D1%84%D0%BE%D1%82%D0%BE%D0%B3%D1%80%D0%B0%D1%84%D0%B8%D0%B8)

alex-ks replied:
Нет, там путаница с названиями запросов. RecognitionTask в запросе делится на две сущности - половина в первом запросе, половина во втором.

whyamiroot replied:
@alex-ks А как тогда сделать? Добавить ID модели и убрать ID сессии?

alex-ks replied:
Да, это то, что нужно.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -0,0 +1,51 @@
+{
+    public class RecognitionTaskQueue
+    {
+        private static RecognitionTaskQueue instance;
```
Не хватает volatile.

whyamiroot replied:
@alex-ks А lock делать не нужно при создании инстанса?

alex-ks replied:
Нужно, да.

alex-ks replied:
Вообще, можно сделать создание синглтона через DI.

whyamiroot committed:
Add volatile keyword to BlockingCollection queue

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -0,0 +1,51 @@
+            RecognitionTask task = null;
+            using (logger.BeginScope(nameof(Dequeue)))
+            {
+                var result = queue.TryTake(out task);
```
Достаточно вызова Take - он заблокируется, пока в очереди не появятся объекты (т.е. то, что нам нужно).

alex-ks replied:
А при таком варианте нам с одной стороны, нужно опрашивать очередь, пока не появятся элементы, с другой стороны, непонятно, что значит null - то, что в очереди пусто, или что в ней лежал null.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+            using (logger.BeginScope(nameof(Enqueue)))
+            {
+                var result = queue.TryAdd(task);
```
Аналогично замечанию про Take, с той разницей, что у нашей очереди нет лимита сверху, поэтому вызов Add никогда не заблокируется.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -0,0 +1,51 @@
+        public static RecognitionTaskQueue GetRecognitionTaskQueue()
+        {
+            if (instance == null)
+                instance = new RecognitionTaskQueue(new LoggerFactory());
```
Нужно получить ту ILoggerFactory, которая зарегистрирована в DI.

whyamiroot committed:
Updated RecognitionTask, RecognitionTaskQueue.

RecognitionTaskQueue injected as Singleton, removed session ID from
RecognitionTask model and added classifier model name

whyamiroot committed:
Photo changed to photoId(int)

DVEfremov commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace Comptech.Backend.Service
+{
+    public class RecognitionTaskQueue
```
лучше на все публичные компоненты добавлять коменты по назначению и примеру использования

DVEfremov commented on Comptech.Backend/src/Comptech.Backend.Service/Models/RecognitionTask.cs:
```diff
@@ -0,0 +1,13 @@
+    public class RecognitionTask
+    {
+        [DataMember(Name = "modelName")]
+        public string ModelName { get; set; }
```
code style 
лучше придерживаться стиля при котором каждый оператор - отдельнная строчка

alex-ks replied:
В C# auto-property обычно пишут в одну строку:
https://msdn.microsoft.com/en-us/library/bb384054.aspx
Если get и set имеют тела, то на отдельных строках.

whyamiroot committed:
Comments for public component RecognitionTaskQueue

Added comments with summary and usage example to RecognitionTaskQueue
and public methods

whyamiroot committed:
Merge branch 'master' into issue-14

whyamiroot committed:
Add tests for RecognitionTaskQueue

Add tests for RecognitionTaskQueue methods - Enqueue, Dequeue. They seem
to be passing on local machine. Fixed Enqueue and Dequeue methods to
use Take and Add methods instead of TryAdd and TryTake.

alex-ks commented on Comptech.Backend/src/Comptech.Backend.Service/RecognitionTaskQueue.cs:
```diff
@@ -5,25 +5,56 @@
         public RecognitionTask Dequeue()
         {
             RecognitionTask task;
             queue.TryTake(out task);
```
TryTake не подходит, поскольку он неблокирующийся. Нужно использовать просто Take.
То же касается TryAdd ниже.

whyamiroot replied:
@alex-ks https://github.com/alex-ks/backend-comptech/pull/23/commits/bf70952a71bab152f12e3d4236fd8fa54391a41b

whyamiroot replied:
@alex-ks сделал ещё пару фиксов. https://github.com/alex-ks/backend-comptech/pull/23/commits/56b57c66bc93a03de60a2821e99772eb2da69b8e

whyamiroot committed:
Fixed testing int against null in Enqueue test.

Added method to ensure queue is empty before testing Enqueue and Dequeue
methods.

whyamiroot committed:
Added information logging in Enqueue and Dequeue
